### PR TITLE
OCL_FP16 MatMul with large batch

### PR DESCRIPTION
**Pull request by dkurt**
### Pull Request Readiness Checklist

* resolves https://github.com/opencv/opencv/issues/24111
* Updates required for actual models from https://github.com/opencv/opencv_extra/pull/1080

**Merge with extra**: https://github.com/opencv/opencv_extra/pull/1080

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Linux OpenCL,Win64 OpenCL
buildworker:Linux OpenCL=linux-1
```

### CODE REVIEW

File: modules/dnn/src/opencl/gemm_buffer.cl [hunk 1]
```diff
@@ -453,14 +453,14 @@ __kernel void TEMPLATE(gemm_buffer_NT, Dtype)(
...
-        vstore4(vload4(0, (__global float *)(src1_read0 + mad24(7, K, local_index))), 0, (__local float *)(slm_brow + mad24(7, SLM_BLOCK, local_index)));
+        vstore8(vload8(0, src1_read0 + mad24(0, K, local_index)), 0, slm_brow + mad24(0, SLM_BLOCK, local_index));
+        vstore8(vload8(0, src1_read0 + mad24(1, K, local_index)), 0, slm_brow + mad24(1, SLM_BLOCK, local_index));
+        vstore8(vload8(0, src1_read0 + mad24(2, K, local_index)), 0, slm_brow + mad24(2, SLM_BLOCK, local_index));
+        vstore8(vload8(0, src1_read0 + mad24(3, K, local_index)), 0, slm_brow + mad24(3, SLM_BLOCK, local_index));
+        vstore8(vload8(0, src1_read0 + mad24(4, K, local_index)), 0, slm_brow + mad24(4, SLM_BLOCK, local_index));
+        vstore8(vload8(0, src1_read0 + mad24(5, K, local_index)), 0, slm_brow + mad24(5, SLM_BLOCK, local_index));
+        vstore8(vload8(0, src1_read0 + mad24(6, K, local_index)), 0, slm_brow + mad24(6, SLM_BLOCK, local_index));
+        vstore8(vload8(0, src1_read0 + mad24(7, K, local_index)), 0, slm_brow + mad24(7, SLM_BLOCK, local_index));
```
Comment (dkurt) [inline modules/dnn/src/opencl/gemm_buffer.cl:463]: Not an OpenCL expert, but not sure why here and below we use `as_half8(vload4(0, (__local float *)...` instead just `vload8` when method is defined for half type only: `#if TYPE == TYPE_HALF`

File: modules/dnn/src/opencl/gemm_buffer.cl [hunk 2]
```diff
@@ -510,7 +510,7 @@ __kernel void TEMPLATE(gemm_buffer_NT, Dtype)(
         Dtype8 arow;
 
 #define READ_BROW(_brow, _row) \
-        _brow = as_half8(vload4(0, (__local float *)(slm_brow0 + _row * SLM_BLOCK))); \
+        _brow = vload8(0, slm_brow0 + _row * SLM_BLOCK); \
```
Comment (dkurt) [inline modules/dnn/src/opencl/gemm_buffer.cl:513]: Current test is `24x5` by `6x5`. Need to add a test with bigger dimensions to go to an inner loop.

File: modules/dnn/test/test_halide_layers.cpp [hunk 3]
```diff
@@ -467,6 +468,7 @@ TEST_P(FullyConnected, Accuracy)
 }
 
 INSTANTIATE_TEST_CASE_P(Layer_Test_Halide, FullyConnected, Combine(
+/*batch*/        Values(1, 2, 4, 8, 24),
```
Comment (dkurt) [inline modules/dnn/test/test_halide_layers.cpp:471]: 24 batch size hangs locally but passed on CI

File: modules/dnn/src/opencl/gemm_buffer.cl (hunk 1 above)
Comment (opencv-alalek) [reply modules/dnn/src/opencl/gemm_buffer.cl:463]: Agreed.

`float` cast hack violates alignment requirement of vload/vstore functions.

Comment (dkurt) [reply modules/dnn/src/opencl/gemm_buffer.cl:463]: However using existing tests I could not catch an error inside the `while( w + TILE_K <= end_w )` loop. Should be changed in the same way?
```cl
brow0 = as_half8(vload4(0, (__local float *)(slm_brow0 + 0 * SLM_BLOCK)));
```
```cl
arow = as_half8(vload4(0, (__global float *)(src0_read + _row * K)));
```

Comment (opencv-alalek) [reply modules/dnn/src/opencl/gemm_buffer.cl:463]: Yes, it is better to change it anywhere.

**Review by opencv-alalek — APPROVED** [review]
Thank you :+1:

**Pull request merged**